### PR TITLE
WIP Feature: More efficient syncing

### PR DESCRIPTION
**Pull request by krconv**
# Description 

Reduces the weight of syncing my only syncing families and surveys that have been created since the last sync time.

# Changes 

- Uses the last sync time to get newly updated families and surveys
  - Sync process adds the last sync time to the request to the REST server
  - Uses a error window of 1 day to ensure nothing strange happens with missing families/surveys
    - The last sync time is the end of last sync, where request was made at an unknown time before that (assuming here that the syncing process never lasted more than one day)
    - Also helps with avoiding issues related to time being changed on a device since last sync
- Adds a last modified field to family
  - Adds a very rough estimate of the families last modified time (basically is the last synced time)

# Tests 

This code has been tested in the following ways:
- Tasks:
  - [X] Logged in to the application
  - [X] Fill out a snapshot for a new family
  - [X] Fill out a snapshot for an existing family
  - [X] Fill out priorities for a snapshot
  - [X] View a family and it's priorities
- API Level:
  - [ ] API Level 19
  - [X] API Level 22
  - [ ] API Level 25
- Screen Size:
  - [ ] 7" screen size
  - [X] 8" screen size
  - [ ] 10" screen size

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,9 +8,6 @@
 /build
 /captures
 .externalNativeBuild
```
Comment (bhylak) [inline .gitignore]: who keeps doing dis

File: app/src/main/java/org/fundacionparaguaya/advisorapp/models/Family.java [hunk 2]
```diff
@@ -213,7 +230,7 @@ public Builder snapshot(Snapshot snapshot) {
         }
 
         public Family build() {
-            return new Family(id, remoteId, name, code, address, location, member, isActive);
+            return new Family(id, remoteId, name, code, address, location, member, isActive, new Date());
```
Comment (bhylak) [inline app/src/main/java/org/fundacionparaguaya/advisorapp/models/Family.java]: there's no way to get the actual last modified time here, right?

File: app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/FamilyRepository.java [hunk 3]
```diff
@@ -69,18 +78,19 @@ public void saveFamily(Family family) {
...
+        return pullFamilies(lastSync);
     }
 
     void clean() {
         familyDao.deleteAll();
     }
 
-    private boolean pullFamilies() {
+    private boolean pullFamilies(Date lastSync) {
         try {
```
Comment (bhylak) [inline app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/FamilyRepository.java]: I think you should check if the date is null, which would trigger a full sync. Just a thought.

File: app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/SnapshotRepository.java [hunk 4]
```diff
@@ -176,9 +177,9 @@ private boolean pushSnapshots() {
         return success;
     }
 
-    private boolean pullSnapshots() {
+    private boolean pullSnapshots(Date lastSync) {
         boolean success = true;
-        for (Family family : familyRepository.getFamiliesNow()) {
+        for (Family family : familyRepository.getFamiliesModifiedSinceDateNow(lastSync)) {
```
Comment (bhylak) [inline app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/SnapshotRepository.java]: function naming is a little weird here "sinceDateNow" 

I know the convention you have set up, so if there's no better wording it's fine.

**Comment by bhylak** [comment]
Fixes #321

**Comment by krconv** [comment]
Not going to merge yet; need to better define the two date fields that I use (last modified, last sync), including when the values should be null and applicable checks

File: app/src/main/java/org/fundacionparaguaya/advisorapp/models/Family.java (hunk 2 above)
Comment (krconv) [reply app/src/main/java/org/fundacionparaguaya/advisorapp/models/Family.java]: Will consider a better way to do this, including scenarios where this builder could possibly be used

File: .gitignore (hunk 1 above)
Comment (krconv) [reply .gitignore]: @alextac98 :)

File: app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/FamilyRepository.java (hunk 3 above)
Comment (krconv) [reply app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/FamilyRepository.java]: Will reconsider

File: app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/SnapshotRepository.java (hunk 4 above)
Comment (bhylak) [reply app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/SnapshotRepository.java]: It's better than no underscore IMO.

.Net convention is "Async" at the end of asynchronous functions. another thing I've seen in .NET is "beginGetFamilies" vs. getFamilies)

Comment (bhylak) [reply app/src/main/java/org/fundacionparaguaya/advisorapp/repositories/SnapshotRepository.java]: I also really like the idea you mentioned earlier... async functions just being the name with no "get" prefix.

So if this were async, it'd be `FamiliesModifiedSinceDate().observe(x,y)`

**Pull request merged**